detectable/src/main/java/com/synopsys/integration/detectable/detectables/yarn/YarnTransformer.java

```diff
@@ -148,9 +149,20 @@ private void collectYarnDependencies(
             ) throws MissingExternalIdException {
         for (YarnLockDependency dependency : entry.getDependencies()) {
             if (!isWorkspace(yarnLockResult.getWorkspaceData(), dependency)) {
+                Map<String, String> idVersionMap = resolvedEntryIdVersionMap.get(dependency.getName());
                 String dependencyVersion;
-                if (resolvedEntryIdVersionMap.containsKey(dependency.getName()) && resolvedEntryIdVersionMap.get(dependency.getName()).containsKey(dependency.getVersion())) {
-                    dependencyVersion = entry.getVersion();
+                if (idVersionMap != null) {
+                    dependencyVersion = idVersionMap.get(dependency.getVersion());
+                    if (dependencyVersion == null) {
+                        if (idVersionMap.values().isEmpty()) {
+                            logger.warn("Dependency {} with version definition {} not found in the Yarn map entries {}", dependency.getName(), dependency.getVersion(), idVersionMap.toString());
+                            dependencyVersion = dependency.getVersion();
+                        } else {
+                            // 1. Choose first version.
+                            dependencyVersion = (String) idVersionMap.values().toArray()[0];
+                        }
+                        // 2. Try to auto-resolve to one of the versions.
+                    }
```

Comment:
A best effort 3 step process to determine version to use for the dependency.
1. Map entry found and version found.
2. Map entry found but version unequal.
3. No map entry, we fallback to the definition version. A further improvement is possible here to intelligently resolve the version based on the range character.